server/src/main/java/org/elasticsearch/indices/SystemIndices.java

```diff
@@ -577,22 +568,26 @@ private static boolean overlaps(String pattern1, String pattern2) {
         return Operations.isEmpty(Operations.intersection(a1Automaton, a2Automaton)) == false;
     }
 
-    private static Map<String, Feature> buildSystemIndexDescriptorMap(Map<String, Feature> featuresMap) {
-        final Map<String, Feature> map = Maps.newMapWithExpectedSize(featuresMap.size() + SERVER_SYSTEM_INDEX_DESCRIPTORS.size());
-        map.putAll(featuresMap);
+    private static List<Feature> buildFullFeatureList(List<Feature> features) {
+        final Map<String, Feature> map = Maps.newMapWithExpectedSize(features.size() + SERVER_SYSTEM_INDEX_DESCRIPTORS.size());
+
+        features.forEach(f -> map.put(f.getName(), f));
+
+        List<String> providedFeatureNames = features.stream().map(Feature::getName).toList();
+
         // put the server items last since we expect less of them
-        SERVER_SYSTEM_INDEX_DESCRIPTORS.forEach((source, feature) -> {
-            if (map.putIfAbsent(source, feature) != null) {
+        SERVER_SYSTEM_INDEX_DESCRIPTORS.forEach(feature -> {
+            if (providedFeatureNames.contains(feature.getName())) {
```

Comment:
This is going to be linear complexity. It looks like this method is only called during initialization, so maybe it is ok to still pass in a Map, and then reduce that to a list of features later cone all the building/validation is complete?

Reply:
I've reverted to the old code that uses the `Map#put` method to detect feature name clashes.